From 3df9be9377ba50b47ea475a316323c45d0c88e06 Mon Sep 17 00:00:00 2001
From: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>
Date: Fri, 9 Feb 2024 15:27:23 -0800
Subject: [PATCH] wifi: ath12k: Update the incorrect wmi channel width value
 sent to fw for 320MHz BW

Currently host is sending the channel width value as 4 instead of 9 for
bandwidth expansion to 320MHz. This is caused due to the difference in
enums followed to set during normal association case and channel width
change case.

Fixed by adding a translation function to handle this case.

Signed-off-by: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index 97c649c..1208d2c 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -556,6 +556,35 @@ ath12k_wmi_get_host_chan_width(u32 width)
 	return host_width;
 }
 
+enum wmi_host_channel_width ath12k_wmi_get_host_chan_switch_width(u32 width)
+{
+	enum wmi_host_channel_width host_width;
+
+	switch (width) {
+	case IEEE80211_STA_RX_BW_20:
+		host_width = WMI_HOST_CHAN_WIDTH_20;
+		break;
+	case IEEE80211_STA_RX_BW_40:
+		host_width = WMI_HOST_CHAN_WIDTH_40;
+		break;
+	case IEEE80211_STA_RX_BW_80:
+		host_width = WMI_HOST_CHAN_WIDTH_80;
+		break;
+	case IEEE80211_STA_RX_BW_160:
+		host_width = WMI_HOST_CHAN_WIDTH_160;
+		break;
+	case IEEE80211_STA_RX_BW_320:
+		host_width = WMI_HOST_CHAN_WIDTH_320;
+		break;
+	default:
+		ath12k_warn(NULL, "invalid bandwidth %d switching back to 20 MHz\n", width);
+		host_width = WMI_HOST_CHAN_WIDTH_20;
+		break;
+	}
+
+	return host_width;
+}
+
 static int
 ath12k_wmi_tlv_iter(struct ath12k_base *ab, const void *ptr, size_t len,
 		    int (*iter)(struct ath12k_base *ab, u16 tag, u16 len,
@@ -15279,6 +15308,7 @@ static void ath12k_wmi_put_peer_list(struct ath12k_base *ab,
 	struct wmi_chan_width_peer_list *itr;
 	struct wmi_chan_width_peer_arg *arg_itr;
 	int i;
+	u32 host_chan_width;
 
 	ath12k_dbg(ab, ATH12K_DBG_WMI,
 		   "wmi peer channel width switch command peer list\n");
@@ -15290,13 +15320,14 @@ static void ath12k_wmi_put_peer_list(struct ath12k_base *ab,
 		itr->tlv_header = ath12k_wmi_tlv_cmd_hdr(WMI_TAG_CHAN_WIDTH_PEER_LIST,
 							 sizeof(*itr));
 		ether_addr_copy(itr->mac_addr.addr, arg_itr->mac_addr.addr);
-		itr->chan_width = cpu_to_le32(arg_itr->chan_width);
+		host_chan_width = ath12k_wmi_get_host_chan_switch_width(arg_itr->chan_width);
+		itr->chan_width = cpu_to_le32(host_chan_width);
 		itr->puncture_20mhz_bitmap = cpu_to_le32(arg_itr->puncture_20mhz_bitmap);
 
 		ath12k_dbg(ab, ATH12K_DBG_WMI,
-			   "   (%u) width %u addr %pM punct_bitmap 0x%x\n",
+			   "   (%u) width %u addr %pM punct_bitmap 0x%x host chan_width: %d\n",
 			   i + 1, arg_itr->chan_width, arg_itr->mac_addr.addr,
-			   arg_itr->puncture_20mhz_bitmap);
+			   arg_itr->puncture_20mhz_bitmap, host_chan_width);
 	}
 }
 
-- 
2.25.1

